### ISSUE TITLE
bug(log): `--include` and `--exclude` flags don't work for `show` and `log`

### ISSUE THREAD

**Issue by vegerot** [facebook/sapling#582]
Steps to reproduce:
1. edit and commit file `x`
2. run `sl log -vp . --include=x`

Expected:
outputs the diff for file `x`

Actual:
No output

**Comment by vegerot** [comment]
I looked into it and `readfiles` in `changelog.py` is not getting the list of changed files in the commit text.

Looking deeper, the implementation of `ReadCommitText` for `git2::Repository` is skipping fetching the files https://github.com/facebook/sapling/blob/49d5c4f0962366ed85f7afe5bb27fc7510632ffe/eden/scm/lib/hgcommits/src/git.rs#L452-L454 .

I assume this is because in Facebook Hg, getting the list of files is fast but is slow in git.

I would like to work on this issue if possible, but I'd like some design help from you guys to determine the best way to do this.  The ideas I have in mind will require some rearchitecting of how we currently resolve commit info

**Comment by vegerot** [comment]
This bug may be related to #581

**Comment by vegerot** [comment]
ping 🔔
I'd like to work on this.  The "easy" solution would be to compute the files changed inside `hgcommits::to_hg_text`.  However, this would be very slow because each commit has to communicate between Python.

What would be faster is to implement the entire filematcher in Rust.  At least this would give Sapling the same performance as `git log --filename`

**Comment by quark-zju** [comment]
> However, this would be very slow because each commit has to communicate between Python.

Right, the performance penalty is undesirable. We want to avoid calculating it unless requested.

In 484e37311ee84258b0058d1d02841652a5afbb59 we changed `ctx.files()` to ignore the empty file list and use the slow path instead. But that didn't cover all the paths. The `file()` revset uses `changelog.readfiles` which hasn't been updated to use a different code path. Perhaps a straightforward fix would be updating the `file()` revset to be aware of Git, and use a different code path.
